### PR TITLE
Accounting oracle tests

### PR DESCRIPTION
**Pull request by vp4242**
Accounting oracle tests

https://github.com/lidofinance/lido-dao/blob/master/test/0.8.9/oracle/accounting-oracle-access-control.test.js
https://github.com/lidofinance/lido-dao/blob/master/test/0.8.9/oracle/accounting-oracle-deploy.test.js
https://github.com/lidofinance/lido-dao/blob/master/test/0.8.9/oracle/accounting-oracle-happy-path.test.js
https://github.com/lidofinance/lido-dao/blob/master/test/0.8.9/oracle/accounting-oracle-submit-report-data.test.js
https://github.com/lidofinance/lido-dao/blob/master/test/0.8.9/oracle/accounting-oracle-submit-report-extra-data.test.js

### CODE REVIEW

File: test/0.8.9/oracle/accountingOracleAccessControl.test.ts [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+      expect(oracle).to.be.not.null;
+      expect(consensus).to.be.not.null;
+      expect(mockLido).to.be.not.null;
+      expect(reportItems).to.be.not.null;
+      expect(extraDataList).to.be.not.null;
+    });
+  });
+
+  context("SUBMIT_DATA_ROLE", () => {
+    beforeEach(deploy);
```
Comment (mymphe) [inline test/0.8.9/oracle/accountingOracleAccessControl.test.ts:105]: Why don't we just `beforeEach` once in the outer context? If we need to use different `extraData`, we can just submit a new report inside the inner context with the necessary `extraData`?

**Review by mymphe — CHANGES_REQUESTED** [review]

File: lib/locator-deploy.ts [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+) {
+  const proxy = await ethers.getContractAt("OssifiableProxy", proxyAddress);
+
+  const contractFactory = await ethers.getContractFactory(artifactName);
+  const implementation = await contractFactory.deploy(...constructorArgs, { from: proxyOwner });
+
+  await proxy.proxy__upgradeTo(await implementation.getAddress());
+}
+
+async function getLocatorConfig(locatorAddress: string) {
```
[the remark this answers is not in the document]
Comment (vp4242) [reply lib/locator-deploy.ts:48]: It seems, I can't. dummyLocator is using LidoLocator Mutable Mock but here it's LidoLocator expected.

File: lib/string.ts [hunk 3]
```diff
@@ -13,3 +13,8 @@ export function en0x(value: number | bigint) {
 export function randomString(length: number) {
   return hexlify(randomBytes(length));
 }
+
+export function hex(n: number, byteLen: number | undefined = undefined) {
```
[the remark this answers is not in the document]
Comment (vp4242) [reply lib/string.ts:17]: bigintToHex() works with bigints that requires conversion from number to bigint. Looks like an overhead

File: test/0.8.9/oracle/accountingOracleSubmitReportExtraData.test.ts [hunk 4]
```diff
@@ -0,0 +1,667 @@
...
+    reportFields?: object;
+  }
+
+  async function takeSnapshot() {
+    snapshot = await Snapshot.take();
+  }
+
+  async function rollback() {
+    await Snapshot.restore(snapshot);
+  }
```
[the remark this answers is not in the document]
Comment (vp4242) [reply test/0.8.9/oracle/accountingOracleSubmitReportExtraData.test.ts:96]: why?

File: test/0.8.9/oracle/accountingOracleAccessControl.test.ts [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+      it("should allow calling from a possessor of SUBMIT_DATA_ROLE role", async () => {
+        const submitDataRole = await oracle.SUBMIT_DATA_ROLE();
+        await oracle.grantRole(submitDataRole, account2);
+        const deadline = (await oracle.getConsensusReport()).processingDeadlineTime;
+        await consensus.setTime(deadline);
+
+        await oracle.connect(account2).submitReportData(reportFields, CONSENSUS_VERSION);
+        const tx = await oracle.connect(account2).submitReportExtraDataList(extraDataList);
+
+        await expect(tx).to.emit(oracle, "ExtraDataSubmitted").withArgs(reportFields.refSlot, anyValue, anyValue);
```
[the remark this answers is not in the document]
Comment (vp4242) [reply test/0.8.9/oracle/accountingOracleAccessControl.test.ts:149]: yes but isn't that more readable this way?

File: test/0.8.9/oracle/accountingOracleDeploy.test.ts [hunk 6]
```diff
@@ -0,0 +1,487 @@
...
+      const deployed = await deployAccountingOracleSetup(admin.address, {
+        getLegacyOracle: () => deployMockLegacyOracle({ epochsPerFrame: EPOCHS_PER_FRAME - 1 }),
+      });
+      await updateInitialEpoch(deployed.consensus);
+      await expect(initAccountingOracle({ admin: admin.address, ...deployed }))
+        .to.be.revertedWithCustomError(deployed.oracle, "IncorrectOracleMigration")
+        .withArgs(1);
+    });
+
+    it(`init fails if the initial epoch of the new oracle is not the next frame's first epoch`, async () => {
```
[the remark this answers is not in the document]
Comment (vp4242) [reply test/0.8.9/oracle/accountingOracleDeploy.test.ts:331]: I'm not the author of the original tests but they make sense to me. Could you elaborate on this particular case?

File: lib/string.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (vp4242) [reply lib/string.ts]: I did the implementation for numberToHex() with bigintToHex() but can't say I like it this way. wdyt?

File: lib/dummy.ts [hunk 7]
```diff
@@ -30,6 +30,67 @@ export async function dummyLocator(config?: Partial<LidoLocator.ConfigStruct>, d
...
 }
+
+export async function dummyLocator(config?: Partial<LidoLocator.ConfigStruct>, deployer?: HardhatEthersSigner) {
+  if (!deployer) {
+    [deployer] = await ethers.getSigners();
+  }
+  const locator = await deployLocator(config, deployer);
+  const proxyFactory = new OssifiableProxy__factory(deployer);
+  const proxy = await proxyFactory.deploy(await locator.getAddress(), await deployer.getAddress(), new Uint8Array());
+  return await ethers.getContractAt("LidoLocator", await proxy.getAddress());
```
[the remark this answers is not in the document]
Comment (mymphe) [reply lib/dummy.ts:43]: Also, I don't think using contact names to refer to the build artifact is a good pattern. What do you think about using `attach`? It returns a new instance of the contract with the same ABI but at a different address.

```suggestion
  return locator.attach(proxy.getAddress());
```

More
https://docs.ethers.org/v6/api/contract/#BaseContract-attach

File: lib/dummy.ts [hunk 8]
```diff
@@ -30,6 +30,67 @@ export async function dummyLocator(config?: Partial<LidoLocator.ConfigStruct>, d
...
+async function updateProxyImplementation(
+  proxyAddress: string,
+  config: LidoLocator.ConfigStruct,
+  customLocator?: string,
+  proxyOwner?: HardhatEthersSigner,
+) {
+  if (!proxyOwner) {
+    [proxyOwner] = await ethers.getSigners();
+  }
+  const proxy = await ethers.getContractAt("OssifiableProxy", proxyAddress);
```
Comment (mymphe) [inline lib/dummy.ts:55]: Here as well. It's better to use `typechain` factories rather than strings to refer to artifacts.

File: lib/dummy.ts [hunk 9]
```diff
@@ -30,6 +30,67 @@ export async function dummyLocator(config?: Partial<LidoLocator.ConfigStruct>, d
...
+    oracleReportSanityChecker: await locator.oracleReportSanityChecker(),
+    postTokenRebaseReceiver: await locator.postTokenRebaseReceiver(),
+    burner: await locator.burner(),
+    stakingRouter: await locator.stakingRouter(),
+    treasury: await locator.treasury(),
+    validatorsExitBusOracle: await locator.validatorsExitBusOracle(),
+    withdrawalQueue: await locator.withdrawalQueue(),
+    withdrawalVault: await locator.withdrawalVault(),
+    oracleDaemonConfig: await locator.oracleDaemonConfig(),
+  };
```
Comment (mymphe) [inline lib/dummy.ts:83]: A small optimization that I would suggest is to parallelize these requests using `Promise.all`.

**Review by mymphe — APPROVED** [review]

File: lib/dummy.ts [hunk 10]
```diff
@@ -30,6 +30,100 @@ export async function dummyLocator(config?: Partial<LidoLocator.ConfigStruct>, d
...
+    await locator.lido(),
+    await locator.oracleReportSanityChecker(),
+    await locator.postTokenRebaseReceiver(),
+    await locator.burner(),
+    await locator.stakingRouter(),
+    await locator.treasury(),
+    await locator.validatorsExitBusOracle(),
+    await locator.withdrawalQueue(),
+    await locator.withdrawalVault(),
+    await locator.oracleDaemonConfig(),
```
Comment (mymphe) [inline lib/dummy.ts:98]: Don't need to `await`. `Promise.all` accepts promises

**Pull request merged**